odl/trafos/convolution.py

```diff
@@ -0,0 +1,804 @@
+            self._transform = None
+
+        scale = kwargs.pop('scale', True)
+
+        # TODO: handle case if axes are given, but the kernel is the same
+        # for each point along the other axes
+        if ker_mode == 'real':
+            # Kernel given as real space element
+            if use_own_ft:
+                self._kernel = None
+                self._kernel_transform = self.transform(kernel, **kwargs)
+                if scale:
+                    self._kernel_transform *= (np.sqrt(2 * np.pi) **
+                                               self.domain.ndim)
+            else:
+                if not self.domain.partition.is_regular:
+                    raise NotImplementedError(
+                        'real-space convolution not implemented for '
+                        'irregular sampling.')
+                try:
+                    # Function or other element-like as input. All axes
+                    # must be used, otherwise we get an error.
+                    # TODO: make a zero-centered space by default and
+                    # adapt the range otherwise
+                    self._kernel = self._kernel_elem(
+                        self.domain.element(kernel).asarray(), self.axes)
+                except (TypeError, ValueError):
+                    # Got an array-like, axes can be used
+                    self._kernel = self._kernel_elem(kernel, self.axes)
+                finally:
+                    self._kernel_transform = None
+                    if scale:
+                        self._kernel *= self.domain.cell_volume
+        else:
+            # Kernel given as Fourier space element
+            self._kernel = None
+            self._kernel_transform = self.transform.range.element(kernel)
+            if scale:
+                self._kernel_transform *= (np.sqrt(2 * np.pi) **
+                                           self.domain.ndim)
+
+    def _kernel_elem(self, kernel, axes):
+        """Return kernel with adapted shape for real space convolution."""
+        kernel = np.asarray(kernel)
+        extra_dims = self.domain.ndim - kernel.ndim
+
+        if extra_dims == 0:
+            return self.domain.element(kernel)
+        else:
+            if len(axes) != extra_dims:
+                raise ValueError('kernel dim {} + number of axes {} '
+                                 '!= space dimension {}.'
+                                 ''.format(kernel.ndim, len(axes),
+                                           self.domain.ndim))
+
+            # Sparse kernel (less dimensions), blow up
+            slc = [None] * self.domain.ndim
+            for ax in axes:
+                slc[ax] = slice(None)
+
+            kernel = kernel[slc]
+            # Assuming uniform discretization
+            min_corner = -self.domain.cell_sides * self.domain.shape / 2
+            max_corner = self.domain.cell_sides * self.domain.shape / 2
+            if isinstance(self.domain.dspace, Ntuples):
```

Comment:
This shall clearly be changed to `impl = self.domain.impl` once the odlcuda stuff is merged.